src/coreclr/jit/lclvars.cpp

```diff
@@ -1790,6 +1788,62 @@ bool Compiler::StructPromotionHelper::CanPromoteStructType(CORINFO_CLASS_HANDLE
     return true;
 }
 
+//--------------------------------------------------------------------------------------------
+// CanConstructAndPromoteField - checks if we can construct field types without asking about them directly.
+//
+// Arguments:
+//   structPromotionInfo - struct promotion candidate information.
+//
+// Return value:
+//   true if we can figure out the fields from available knowledge.
+//
+// Notes:
+//   This is needed for AOT R2R compilation when we can't cross compilation bubble borders
+//   so we should not ask about fields that are not directly referenced. If we do VM will have
+//   to emit a type check for this field type but it does not have enough information about it.
+//   As a workaround for perfomance critical corner case: struct with 1 gcref, we try to construct
+//   the field information from indirect observations.
+//
+bool Compiler::StructPromotionHelper::CanConstructAndPromoteField(lvaStructPromotionInfo* structPromotionInfo)
+{
+    const CORINFO_CLASS_HANDLE typeHnd    = structPromotionInfo->typeHnd;
+    const COMP_HANDLE          compHandle = compiler->info.compCompHnd;
+    const DWORD                typeFlags  = compHandle->getClassAttribs(typeHnd);
+    if (structPromotionInfo->fieldCnt != 1)
+    {
+        // Can't find out values for several fields.
+        return false;
+    }
+    if ((typeFlags & CORINFO_FLG_CONTAINS_GC_PTR) == 0)
+    {
+        // Can't find out type of a non-gc field.
+        return false;
+    }
+
+    const unsigned structSize = compHandle->getClassSize(typeHnd);
```

Comment:
This is probably worth a comment as this exact line has key importance for version resiliency. In the JIT interface in <code>CorInfoImpl</code>, the method <code>getClassSize</code> is the one that implicitly triggers creation of the <code>ENCODE_TYPE_LAYOUT_CHECK</code> fixup that is in turn required to invalidate the pre-jitted code when the shape of the struct in a separately versioned module changes (say, when in your example someone changes the type of the <code>o</code> field from Object to IntPtr or adds a new field <code>n</code> before the field <code>o</code> without recompiling the code in which the promotion has taken place).

Reply:
Thanks @trylek. 
Just to confirm, lets say we have 
```
struct OuterStruct {
  InnerClass innerField;
}
```
I can't think of a scenario where before my changes we did not have a `ENCODE_TYPE_LAYOUT_CHECK ` for `OuterStruct ` and after my changes we have them. We also don't have a check for `InnerClass ` before and after the changes. If the jit was working with `OuterStruct` it would have to know its size and ask if there are byref fields for all operations: initialization, copying, passing as arguments and returning. So if we see a struct the jit will always ask these questions so asking them here does not change anything. 

There could be a scenario where we prove that the struct is unused and can be deleted but I believe it happens during liveness when all these questions are already asked to create right copy/init tree nodes.

So I don't think that this line is important and the comment only here would be confusing for me, like why all other places  where we call `getClassSize` don't have such comment.

Reply:
OK, you have convinced me that the code comment I suggested wouldn't have much value. I guess I'm mostly somewhat concerned about the fact that <code>getClassAttrib</code> doesn't emit the type layout check, I'd love to hear from David whether that's intentional (there's no reasonable scenario where JIT would only ask about the attribs and not size), opportunistic (perhaps an additional place emitting the fixup might somewhat slow down the compilation) or just an omission. One way or another, I have already approved your change and on my part I don't see anything preventing it from being merged in.